### DISCUSSION TITLE
Some clarification questions for benchmarking

### DISCUSSION THREAD

**Discussion by shy218** [sgkit-dev/sgkit#528] [General]
I have a few concerns about this research and our goals.
1. What are the packages we want to compare? I currently select the SGKIT, Hail, Glow, Plink, pbdR and SystemML. pbdR and SystemML are two low-level packages, which are not specific to genomics datasets. If we add these two packages, I need to reimplement all the algorithms again from scratch, and the workload is huge. This is not doable in just two months.
2. And also, there is another related question. What is our input format? VCF file or genotype call? If our input format is VCF file, do we need to also write the functions to convert VCF to genotype call? Or could we call the functions from SGKIT? If our input format is genotype call, some packages, such as Plink and Glow cannot be tested. I currently prefer to test on some packages with the same input formats and we select some functions that all the packages have.
3. If I need to implement the algorithms on my own, my implementation may be biased. I think our goal is to test and benchmark the performance of the packages for genomics datasets, not to reimplement and try to optimize the algorithm. I am not an expert in Bioinfo, and maybe some implementation by myself is not optimized.
Sorry for bothering you time. I am confused after the deep analysis of the packages and algorithms this week.

**Comment by hammer** [comment]
> What are the packages we want to compare?

I think `sgkit`, Hail, and PLINK are fine to start.

> What is our input format?

That's a good question. I think we should use VCF as the starting point and allow for a conversion step for each library into the library's preferred format, with subsequent operations making use of that preferred format.

> If I need to implement the algorithms on my own

You should not need to implement any algorithms on your own.

**Reply by shy218** [reply]
Thanks for the reply. After analysis on packages and algorithms, I prefer to test Alleles Frequency, Call rate, HWE, LD-prune with windowing, PCA (eig decomposition in Plink and Hail, SVD in Sgkit), OLS with multiple covariates, these six algorithms on Sgkit, Hail and Plink. I think these six algorithms are classical in genetics analysis, and they have fully implementations in all three packages. We will take VCF as input format, as all the packages have functions to read VCF (Plink 1.90 can read vcf.)

**Reply by eric-czech** [reply]
That's great @shy218!  

As far as I know, PLINK 2.0 has been in "alpha" mode for 4+ years and isn't used widely so making comparisons to PLINK 1.9 is good (it's tried and true).

For HWE, are you going to include the per-variant tallying of homozygous reference, heterozygous, and homozygous alternate allele counts (across the samples) with the benchmark?  PLINK for example wouldn't let you feed those counts in directly -- it would force you to calculate the tallies as well while Hail can work either way.  There might be some other incongruities in the comparisons like that but let us know if you have any questions.

Thanks for taking a deeper look.

**Reply by shy218** [reply]
Thanks for the reply. My current plan is to just call the functions in each package. Plink is like a black box to me, and I can't see its inside implementation, so I just call the HWE functions on the datasets and see the result and runtimes. Maybe there are some differences in implementation, but we can try this at first, and then discuss the result.

**Reply by hammer** [reply]
> I prefer to test Alleles Frequency, Call rate, HWE, LD-prune with windowing, PCA (eig decomposition in Plink and Hail, SVD in Sgkit), OLS with multiple covariates

These methods sound like a great start! Let me try to map them to @eric-czech's [categories](https://github.com/pystatgen/sgkit/discussions/481#discussioncomment-430892):
1. Summary statistics for QC: Alleles Frequency, Call rate, HWE
2. LD estimation / pruning: LD-prune with windowing
3. Relatedness estimation / pruning: N/A
4. Variant normalization: N/A
5. Population structure estimation: PCA/SVD
6. Association testing: OLS with multiple covariates

That looks like a great start. @eric-czech any thoughts on if/how we should benchmark relatedness estimation and pruning, and if/how we can include @aktech's great work on pairwise distance computation?

PLINK 1.9 has [--make-rel](https://www.cog-genomics.org/plink/1.9/distance#make_rel) and [-rel-cutoff](https://www.cog-genomics.org/plink/1.9/distance#rel_cutoff), and PLINK 2.0 has [--make-king and --king-cutoff](https://www.cog-genomics.org/plink/2.0/distance#make_king). Hail has both [KING and PC-Relate](https://hail.is/docs/0.2/methods/relatedness.html). Should we try to compare to them somehow?

**Reply by hammer** [reply]
> Plink is like a black box to me

If you mean the code is hard to understand, that's fair! In case you mean you don't know where to find the code, it's at https://github.com/chrchang/plink-ng.

**Reply by eric-czech** [reply]
> Hail has both KING and PC-Relate. Should we try to compare to them somehow?

Good point, I think an sgkit.pc_relate comparison to Hail's [pc_relate](https://hail.is/docs/0.2/methods/relatedness.html#hail.methods.pc_relate) would be excellent.  @ravwojdyla has good examples for that in https://github.com/related-sciences/gwas-analysis/blob/rav-pc_relate-tw/notebooks/platform/xarray/pc_relate_perf.py#L3-L7 and perhaps elsewhere?

> @eric-czech any thoughts on if/how we should benchmark relatedness estimation and pruning

I don't see much value in doing relatedness pruning since making a representative choice from such small groups is easy by comparison to the same problem with LD.  In other words, the family clusters are much easier to isolate than LD clusters and pruning AFAIK typically means just picking someone from a cluster with some ad-hoc routine.

> @aktech's great work on pairwise distance computation?

I think the most natural target for this is `https://hail.is/docs/0.2/methods/stats.html#hail.methods.row_correlation`, which is the same as computing a dense LD matrix for specific inputs (i.e. pearson correlation between variants).  That would be a nice one to include since it's probably more applicable to popgen work as well.

**Reply by ravwojdyla** [reply]
Regarding PC-Relate comparison with Hail, this is a good start: https://discourse.pystatgen.org/t/pc-relate-experiment/51, there are further links in that post.

Also regarding comparison with Hail, as pointed out in https://github.com/related-sciences/gwas-analysis/issues/4#issuecomment-756297291:

>  it would be interesting to see a perf difference between numpy (and/or Dask Array) and Hail's BlockMatrix.

Hail's [BlockMatrix](https://hail.is/docs/0.2/linalg/hail.linalg.BlockMatrix.html) could be compared with Dask's Array, on linear algebra operators. This could be a "low" level benchmark, that builds up to more use case/genetic specific benchmarks.

**Reply by shy218** [reply]
We can replace "call rate" benchmark to pc-relate, since pc-relate is more interesting to be compared.

**Reply by shy218** [reply]
I think we should have three axis of benchmarks -- Complexity or Difficulty of Algorithm, Data Factors (Data size) and Computational Factors (scalability). These are consistent with the SLAB paper from Professor Arun Kumar.

**Comment by hammer** [comment]
I wonder if these benchmarks are overly focused on statistical genetics. @tomwhite are there any unique computational patterns from the population genetics methods you've implemented that might make sense to add to the benchmark?

**Reply by tomwhite** [reply]
The [Garud H](https://pystatgen.github.io/sgkit/latest/generated/sgkit.Garud_H.html#sgkit.Garud_H) stats are the most computationally expensive of the popgen methods that I ported. This is because entries in the samples dimension are hashed and sorted. I don't think there are Hail or PLINK equivalents though, so there wouldn't be anything to compare to.
